swayfmt/src/formatter/shape.rs

```diff
@@ -103,26 +103,51 @@ impl Sub for Indent {
         Indent::new(self.block_indent - rhs.block_indent)
     }
 }
-
+/// Information about the line of code currently being evaluated.
 #[derive(Copy, Clone, Debug, PartialEq, Eq, Default)]
 pub(crate) struct CodeLine {
+    /// The current number of characters in the given code line.
+    pub(crate) width: usize,
     pub(crate) line_style: LineStyle,
     pub(crate) expr_kind: ExprKind,
+    /// Used in determining `SameLineWhere` formatting.
+    pub(crate) has_where_clause: bool,
 }
 
 impl CodeLine {
+    pub(crate) fn from(line_style: LineStyle, expr_kind: ExprKind) -> Self {
+        Self {
+            width: Default::default(),
+            line_style,
+            expr_kind,
+            has_where_clause: Default::default(),
+        }
+    }
+    pub(crate) fn reset_width(&mut self) {
+        self.width = 0;
+    }
+    pub(crate) fn update_width(&mut self, new_width: usize) {
+        self.width = new_width;
+    }
+    pub(crate) fn add_width(&mut self, extra_width: usize) {
+        self.width += extra_width;
+    }
+    pub(crate) fn sub_width(&mut self, extra_width: usize) {
+        self.width -= extra_width;
+    }
     /// Update `CodeLine::line_style` with a given LineStyle.
-    pub(crate) fn update_line_style(&mut self, line_style: LineStyle) {
-        self.line_style = line_style
+    pub(crate) fn with_line_style(&mut self, line_style: LineStyle) {
+        self.line_style = line_style;
     }
     /// Update `CodeLine::expr_kind` with a given ExprKind.
-    pub(crate) fn update_expr_kind(&mut self, expr_kind: ExprKind) {
-        self.expr_kind = expr_kind
+    pub(crate) fn with_expr_kind(&mut self, expr_kind: ExprKind) {
+        self.expr_kind = expr_kind;
     }
-    pub(crate) fn new(line_style: LineStyle, expr_kind: ExprKind) -> Self {
-        Self {
-            line_style,
-            expr_kind,
+    /// Update the value of `has_where_clause`.
+    pub(crate) fn update_where_clause(&mut self) {
```

Comment:
Maybe `toggle_where_clause` could be a clearer name for this?

That said, it might instead be worth passing the value explicitly here (like `update_width`) so that it's clear to the caller what the resulting state will be, as the implementer of `Format` doesn't necessarily know what state the value will be before calling `update_where_clause`.

Reply:
I was thinking that as well, but left it as is since it was working. I'll go ahead and add that change!